jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,256 @@
+        - puzzle: jax array (int32) of shape (N, N), representing the current state of the puzzle.
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - key: jax array (uint32) of shape (2,), random key used to generate random numbers
+            at each step and for auto-reset.
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        viewer: Optional[Viewer[State]] = None,
+    ) -> None:
+        """Instantiate a `SlidingTilePuzzle` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` which generates puzzles with
+                a size of 5x5.
+
+            viewer: environment viewer for rendering.
+        """
+        grid_size = 5
+        # The generator generates a shuffled puzzle
+        self.generator = generator or RandomGenerator(grid_size)
+
+        # Create viewer used for rendering
+        self._env_viewer = viewer or SlidingTilePuzzleViewer(name="SlidingTilePuzzle")
+        self.movements = jnp.array(
+            [[-1, 0], [1, 0], [0, -1], [0, 1]]  # Up  # Down  # Left  # Right
+        )
+        self.solved_puzzle = jnp.arange(grid_size**2).reshape((grid_size, grid_size))
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+        )
+        action_mask = self._get_valid_actions(empty_tile_position)
+        obs = Observation(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            action_mask=action_mask,
+        )
+        timestep = restart(observation=obs)
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Updates the environment state after the agent takes an action."""
+        (updated_puzzle, updated_empty_tile_position) = self._move_empty_tile(
+            state.puzzle, state.empty_tile_position, action
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+
+        # Update the action mask
+        action_mask = self._get_valid_actions(updated_empty_tile_position)
+
+        next_state = State(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            key=state.key,
+        )
+        obs = Observation(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            action_mask=action_mask,
+        )
+
+        reward = -jnp.sum(obs.puzzle != self.solved_puzzle).astype(jnp.float32)
```

Comment:
This reward definition does not mean solving the puzzle asap but rather minimizing the number of wrongly located tiles asap. I would suggest externalizing the reward function (e.g. in TSP) to have it as an argument to the environment constructor. We could have a dense and sparse definition:
- dense: +1 for each newly correctly located tile, -1 for each newly wrongly located tile.
- sparse: +1 if the puzzle is completely solved.